File: tauri/src/midi/pedals/clean/mapper.rs
// Chase Bliss Audio Clean MIDI CC mapping - infrastructure layer

use super::types::{CleanParameter, CleanState};
use std::collections::HashMap;

pub const CC_PRESET_SAVE: u8 = 111;

impl CleanParameter {
    /// Get the CC number for this parameter
    pub fn cc_number(&self) -> u8 {
        match self {
            CleanParameter::Dynamics(_) => 14,
            CleanParameter::Sensitivity(_) => 15,
            CleanParameter::Wet(_) => 16,
            CleanParameter::Attack(_) => 17,
            CleanParameter::Eq(_) => 18,
            CleanParameter::Dry(_) => 19,
            CleanParameter::RampSpeed(_) => 20,

            CleanParameter::ReleaseMode(_) => 21,
            CleanParameter::EffectMode(_) => 22,
            CleanParameter::PhysicsMode(_) => 23,

            CleanParameter::NoiseGateRelease(_) => 24,
            CleanParameter::NoiseGateSens(_) => 25,
            CleanParameter::SwellIn(_) => 26,
            CleanParameter::UserRelease(_) => 27,
            CleanParameter::BalanceFilter(_) => 28,
            CleanParameter::SwellOut(_) => 29,
            CleanParameter::EnvelopeMode(_) => 31,
            CleanParameter::ShiftyMode(_) => 32,
            CleanParameter::SpreadRouting(_) => 33,

            CleanParameter::Bypass(_) => 102,
            CleanParameter::Swell(_) => 103,
            CleanParameter::AltMode(_) => 104,
            CleanParameter::SwellHold(_) => 105,
            CleanParameter::DynamicsMax(_) => 106,

            CleanParameter::DipDynamics(_) => 61,
            CleanParameter::DipAttack(_) => 62,
            CleanParameter::DipEq(_) => 63,
            CleanParameter::DipDry(_) => 64,
            CleanParameter::DipWet(_) => 65,
            CleanParameter::DipBounce(_) => 66,
            CleanParameter::DipSweep(_) => 67,
            CleanParameter::DipPolarity(_) => 68,

            CleanParameter::DipMiso(_) => 71,
            CleanParameter::DipSpread(_) => 72,
            CleanParameter::DipLatch(_) => 73,
            CleanParameter::DipSidechain(_) => 74,
            CleanParameter::DipNoiseGate(_) => 75,
            CleanParameter::DipMotion(_) => 76,
            CleanParameter::DipSwellAux(_) => 77,
            CleanParameter::DipDusty(_) => 78,

            CleanParameter::RampBounce(_) => 52,
            CleanParameter::Expression(_) => 100,
            CleanParameter::PresetSave(_) => 111,
        }
    }

    /// Get the CC value for this parameter
    pub fn cc_value(&self) -> u8 {
        match self {
            // Continuous parameters
            CleanParameter::Dynamics(v)
            | CleanParameter::Sensitivity(v)
            | CleanParameter::Wet(v)
            | CleanParameter::Attack(v)
            | CleanParameter::Eq(v)
            | CleanParameter::Dry(v)
            | CleanParameter::RampSpeed(v)
            | CleanParameter::NoiseGateRelease(v)
            | CleanParameter::NoiseGateSens(v)
            | CleanParameter::SwellIn(v)
            | CleanParameter::UserRelease(v)
            | CleanParameter::BalanceFilter(v)
            | CleanParameter::SwellOut(v)
            | CleanParameter::ShiftyMode(v)
            | CleanParameter::Expression(v)
            | CleanParameter::PresetSave(v) => *v,

            // Enum parameters
            CleanParameter::ReleaseMode(m) => m.to_cc_value(),
            CleanParameter::EffectMode(m) => m.to_cc_value(),
            CleanParameter::PhysicsMode(m) => m.to_cc_value(),
            CleanParameter::EnvelopeMode(m) => m.to_cc_value(),
            CleanParameter::SpreadRouting(m) => m.to_cc_value(),
            CleanParameter::DipSweep(d) => d.to_cc_value(),
            CleanParameter::DipPolarity(p) => p.to_cc_value(),

            // Binary parameters
            CleanParameter::Bypass(b)
            | CleanParameter::Swell(b)
            | CleanParameter::AltMode(b)
            | CleanParameter::SwellHold(b)
            | CleanParameter::DynamicsMax(b)
            | CleanParameter::DipDynamics(b)
            | CleanParameter::DipAttack(b)
            | CleanParameter::DipEq(b)
            | CleanParameter::DipDry(b)
            | CleanParameter::DipWet(b)
            | CleanParameter::DipBounce(b)
            | CleanParameter::DipMiso(b)
            | CleanParameter::DipSpread(b)
            | CleanParameter::DipLatch(b)
            | CleanParameter::DipSidechain(b)
            | CleanParameter::DipNoiseGate(b)
            | CleanParameter::DipMotion(b)
            | CleanParameter::DipSwellAux(b)
            | CleanParameter::DipDusty(b)
            | CleanParameter::RampBounce(b) => {
                if *b { 127 } else { 0 }
            }
        }
    }

    /// Get a human-readable name for this parameter
    pub fn name(&self) -> &'static str {
        match self {
            CleanParameter::Dynamics(_) => "Dynamics",
            CleanParameter::Sensitivity(_) => "Sensitivity",
            CleanParameter::Wet(_) => "Wet",
            CleanParameter::Attack(_) => "Attack",
            CleanParameter::Eq(_) => "EQ",
            CleanParameter::Dry(_) => "Dry",
            CleanParameter::RampSpeed(_) => "Ramp Speed",
            CleanParameter::ReleaseMode(_) => "Release Mode",
            CleanParameter::EffectMode(_) => "Effect Mode",
            CleanParameter::PhysicsMode(_) => "Physics Mode",
            CleanParameter::NoiseGateRelease(_) => "Noise Gate Release",
            CleanParameter::NoiseGateSens(_) => "Noise Gate Sens",
            CleanParameter::SwellIn(_) => "Swell In",
            CleanParameter::UserRelease(_) => "User Release",
            CleanParameter::BalanceFilter(_) => "Balance Filter",
            CleanParameter::SwellOut(_) => "Swell Out",
            CleanParameter::EnvelopeMode(_) => "Envelope Mode",
            CleanParameter::ShiftyMode(_) => "Shifty Mode",
            CleanParameter::SpreadRouting(_) => "Spread Routing",
            CleanParameter::Bypass(_) => "Bypass",
            CleanParameter::Swell(_) => "Swell",
            CleanParameter::AltMode(_) => "Alt Mode",
            CleanParameter::SwellHold(_) => "Swell Hold",
            CleanParameter::DynamicsMax(_) => "Dynamics Max",
            CleanParameter::DipDynamics(_) => "DIP: Dynamics",
            CleanParameter::DipAttack(_) => "DIP: Attack",
            CleanParameter::DipEq(_) => "DIP: EQ",
            CleanParameter::DipDry(_) => "DIP: Dry",
            CleanParameter::DipWet(_) => "DIP: Wet",
            CleanParameter::DipBounce(_) => "DIP: Bounce",
            CleanParameter::DipSweep(_) => "DIP: Sweep",
            CleanParameter::DipPolarity(_) => "DIP: Polarity",
            CleanParameter::DipMiso(_) => "DIP: Miso",
            CleanParameter::DipSpread(_) => "DIP: Spread",
            CleanParameter::DipLatch(_) => "DIP: Latch",
            CleanParameter::DipSidechain(_) => "DIP: Sidechain",
            CleanParameter::DipNoiseGate(_) => "DIP: Noise Gate",
            CleanParameter::DipMotion(_) => "DIP: Motion",
            CleanParameter::DipSwellAux(_) => "DIP: Swell Aux",
            CleanParameter::DipDusty(_) => "DIP: Dusty",
            CleanParameter::RampBounce(_) => "Ramp/Bounce",
            CleanParameter::Expression(_) => "Expression",
            CleanParameter::PresetSave(_) => "Preset Save",
        }
    }
}

impl CleanState {
    /// Convert the current state to a map of CC numbers → CC values.
    /// Used when recalling a full preset (sending all parameters at once).
    pub fn to_cc_map(&self) -> HashMap<u8, u8> {
        let mut map = HashMap::new();

        // Main knobs
        map.insert(14, self.dynamics);
        map.insert(15, self.sensitivity);
        map.insert(16, self.wet);
        map.insert(17, self.attack);
        map.insert(18, self.eq);
        map.insert(19, self.dry);
        map.insert(20, self.ramp_speed);

        // Toggles
        map.insert(21, self.release_mode.to_cc_value());
        map.insert(22, self.effect_mode.to_cc_value());
        map.insert(23, self.physics_mode.to_cc_value());

        // Hidden options
        map.insert(24, self.noise_gate_release);
        map.insert(25, self.noise_gate_sens);
        map.insert(26, self.swell_in);
        map.insert(27, self.user_release);
        map.insert(28, self.balance_filter);
        map.insert(29, self.swell_out);
        map.insert(31, self.envelope_mode.to_cc_value());
        map.insert(32, self.shifty_mode);
        map.insert(33, self.spread_routing.to_cc_value());

        // Footswitches
        map.insert(102, if self.bypass { 127 } else { 0 });
        map.insert(103, if self.swell { 127 } else { 0 });
        map.insert(104, if self.alt_mode { 127 } else { 0 });
        map.insert(105, if self.swell_hold { 127 } else { 0 });
        map.insert(106, if self.dynamics_max { 127 } else { 0 });

        // DIP switches - Left bank
        map.insert(61, if self.dip_dynamics { 127 } else { 0 });
        map.insert(62, if self.dip_attack { 127 } else { 0 });
        map.insert(63, if self.dip_eq { 127 } else { 0 });
        map.insert(64, if self.dip_dry { 127 } else { 0 });
        map.insert(65, if self.dip_wet { 127 } else { 0 });
        map.insert(66, if self.dip_bounce { 127 } else { 0 });
        map.insert(67, self.dip_sweep.to_cc_value());
        map.insert(68, self.dip_polarity.to_cc_value());

        // DIP switches - Right bank
        map.insert(71, if self.dip_miso { 127 } else { 0 });
        map.insert(72, if self.dip_spread { 127 } else { 0 });
        map.insert(73, if self.dip_latch { 127 } else { 0 });
        map.insert(74, if self.dip_sidechain { 127 } else { 0 });
        map.insert(75, if self.dip_noise_gate { 127 } else { 0 });
        map.insert(76, if self.dip_motion { 127 } else { 0 });
        map.insert(77, if self.dip_swell_aux { 127 } else { 0 });
        map.insert(78, if self.dip_dusty { 127 } else { 0 });

        // Utility
        map.insert(52, if self.ramp_bounce { 127 } else { 0 });
        map.insert(100, self.expression);

        map
    }
}

File: apps/web/src/lib/midi/pedals/clean/mapper.ts
// MIDI CC mapping for Clean parameters
// Ported from apps/desktop/tauri/src/midi/pedals/clean/mapper.rs

import type {
  CleanParameter,
  CleanState,
  ReleaseMode,
  EffectMode,
  PhysicsMode,
  EnvelopeMode,
  SpreadRouting,
  SweepDirection,
  Polarity,
} from './types';

// ============================================================================
// Enum → CC value helpers
// ============================================================================

function releaseModeToCCValue(m: ReleaseMode): number {
  switch (m) { case 'Fast': return 1; case 'User': return 2; case 'Slow': return 3; }
}
function releaseModeFromCCValue(v: number): ReleaseMode {
  if (v === 2) return 'User'; if (v === 3) return 'Slow'; return 'Fast';
}

function effectModeToCCValue(m: EffectMode): number {
  switch (m) { case 'Shifty': return 1; case 'Manual': return 2; case 'Modulated': return 3; }
}
function effectModeFromCCValue(v: number): EffectMode {
  if (v === 2) return 'Manual'; if (v === 3) return 'Modulated'; return 'Shifty';
}

function physicsModeToCCValue(m: PhysicsMode): number {
  switch (m) { case 'Wobbly': return 1; case 'Off': return 2; case 'Twitchy': return 3; }
}
function physicsModeFromCCValue(v: number): PhysicsMode {
  if (v === 2) return 'Off'; if (v === 3) return 'Twitchy'; return 'Wobbly';
}

function envelopeModeToCCValue(m: EnvelopeMode): number {
  switch (m) { case 'Analog': return 1; case 'Hybrid': return 2; case 'Adaptive': return 3; }
}
function envelopeModeFromCCValue(v: number): EnvelopeMode {
  if (v === 2) return 'Hybrid'; if (v === 3) return 'Adaptive'; return 'Analog';
}

function spreadRoutingToCCValue(r: SpreadRouting): number {
  switch (r) { case 'Eq': return 1; case 'Both': return 2; case 'VolComp': return 3; }
}
function spreadRoutingFromCCValue(v: number): SpreadRouting {
  if (v === 2) return 'Both'; if (v === 3) return 'VolComp'; return 'Eq';
}

function sweepDirectionToCCValue(d: SweepDirection): number { return d === 'Bottom' ? 0 : 127; }
function sweepDirectionFromCCValue(v: number): SweepDirection { return v < 64 ? 'Bottom' : 'Top'; }
function polarityToCCValue(p: Polarity): number { return p === 'Forward' ? 0 : 127; }
function polarityFromCCValue(v: number): Polarity { return v < 64 ? 'Forward' : 'Reverse'; }

// ============================================================================
// Parameter → CC number
// ============================================================================

export function parameterCCNumber(param: CleanParameter): number {
  if ('Dynamics' in param)        return 14;
  if ('Sensitivity' in param)     return 15;
  if ('Wet' in param)             return 16;
  if ('Attack' in param)          return 17;
  if ('Eq' in param)              return 18;
  if ('Dry' in param)             return 19;
  if ('RampSpeed' in param)       return 20;
  if ('ReleaseMode' in param)     return 21;
  if ('EffectMode' in param)      return 22;
  if ('PhysicsMode' in param)     return 23;
  if ('NoiseGateRelease' in param) return 24;
  if ('NoiseGateSens' in param)   return 25;
  if ('SwellIn' in param)         return 26;
  if ('UserRelease' in param)     return 27;
  if ('BalanceFilter' in param)   return 28;
  if ('SwellOut' in param)        return 29;
  if ('EnvelopeMode' in param)    return 31;
  if ('ShiftyMode' in param)      return 32;
  if ('SpreadRouting' in param)   return 33;
  if ('RampBounce' in param)      return 52;
  if ('DipDynamics' in param)     return 61;
  if ('DipAttack' in param)       return 62;
  if ('DipEq' in param)           return 63;
  if ('DipDry' in param)          return 64;
  if ('DipWet' in param)          return 65;
  if ('DipBounce' in param)       return 66;
  if ('DipSweep' in param)        return 67;
  if ('DipPolarity' in param)     return 68;
  if ('DipMiso' in param)         return 71;
  if ('DipSpread' in param)       return 72;
  if ('DipLatch' in param)        return 73;
  if ('DipSidechain' in param)    return 74;
  if ('DipNoiseGate' in param)    return 75;
  if ('DipMotion' in param)       return 76;
  if ('DipSwellAux' in param)     return 77;
  if ('DipDusty' in param)        return 78;
  if ('Expression' in param)      return 100;
  if ('Bypass' in param)          return 102;
  if ('Swell' in param)           return 103;
  if ('AltMode' in param)         return 104;
  if ('SwellHold' in param)       return 105;
  if ('DynamicsMax' in param)     return 106;
  if ('PresetSave' in param)      return 111;
  throw new Error(`Unknown CleanParameter: ${JSON.stringify(param)}`);
}

// ============================================================================
// Parameter → CC value
// ============================================================================

export function parameterCCValue(param: CleanParameter): number {
  if ('Dynamics' in param)        return param.Dynamics;
  if ('Sensitivity' in param)     return param.Sensitivity;
  if ('Wet' in param)             return param.Wet;
  if ('Attack' in param)          return param.Attack;
  if ('Eq' in param)              return param.Eq;
  if ('Dry' in param)             return param.Dry;
  if ('RampSpeed' in param)       return param.RampSpeed;
  if ('NoiseGateRelease' in param) return param.NoiseGateRelease;
  if ('NoiseGateSens' in param)   return param.NoiseGateSens;
  if ('SwellIn' in param)         return param.SwellIn;
  if ('UserRelease' in param)     return param.UserRelease;
  if ('BalanceFilter' in param)   return param.BalanceFilter;
  if ('SwellOut' in param)        return param.SwellOut;
  if ('ShiftyMode' in param)      return param.ShiftyMode;
  if ('Expression' in param)      return param.Expression;
  if ('PresetSave' in param)      return param.PresetSave;
  if ('ReleaseMode' in param)     return releaseModeToCCValue(param.ReleaseMode);
  if ('EffectMode' in param)      return effectModeToCCValue(param.EffectMode);
  if ('PhysicsMode' in param)     return physicsModeToCCValue(param.PhysicsMode);
  if ('EnvelopeMode' in param)    return envelopeModeToCCValue(param.EnvelopeMode);
  if ('SpreadRouting' in param)   return spreadRoutingToCCValue(param.SpreadRouting);
  if ('DipSweep' in param)        return sweepDirectionToCCValue(param.DipSweep);
  if ('DipPolarity' in param)     return polarityToCCValue(param.DipPolarity);
  // Binary params
  if ('Bypass' in param)          return param.Bypass ? 127 : 0;
  if ('Swell' in param)           return param.Swell ? 127 : 0;
  if ('AltMode' in param)         return param.AltMode ? 127 : 0;
  if ('SwellHold' in param)       return param.SwellHold ? 127 : 0;
  if ('DynamicsMax' in param)     return param.DynamicsMax ? 127 : 0;
  if ('DipDynamics' in param)     return param.DipDynamics ? 127 : 0;
  if ('DipAttack' in param)       return param.DipAttack ? 127 : 0;
  if ('DipEq' in param)           return param.DipEq ? 127 : 0;
  if ('DipDry' in param)          return param.DipDry ? 127 : 0;
  if ('DipWet' in param)          return param.DipWet ? 127 : 0;
  if ('DipBounce' in param)       return param.DipBounce ? 127 : 0;
  if ('DipMiso' in param)         return param.DipMiso ? 127 : 0;
  if ('DipSpread' in param)       return param.DipSpread ? 127 : 0;
  if ('DipLatch' in param)        return param.DipLatch ? 127 : 0;
  if ('DipSidechain' in param)    return param.DipSidechain ? 127 : 0;
  if ('DipNoiseGate' in param)    return param.DipNoiseGate ? 127 : 0;
  if ('DipMotion' in param)       return param.DipMotion ? 127 : 0;
  if ('DipSwellAux' in param)     return param.DipSwellAux ? 127 : 0;
  if ('DipDusty' in param)        return param.DipDusty ? 127 : 0;
  if ('RampBounce' in param)      return param.RampBounce ? 127 : 0;
  throw new Error(`Unknown CleanParameter: ${JSON.stringify(param)}`);
}

// ============================================================================
// State → CC map
// ============================================================================

export function stateToCCMap(state: CleanState): Map<number, number> {
  const map = new Map<number, number>();
  map.set(14, state.dynamics);
  map.set(15, state.sensitivity);
  map.set(16, state.wet);
  map.set(17, state.attack);
  map.set(18, state.eq);
  map.set(19, state.dry);
  map.set(20, state.ramp_speed);
  map.set(21, releaseModeToCCValue(state.release_mode));
  map.set(22, effectModeToCCValue(state.effect_mode));
  map.set(23, physicsModeToCCValue(state.physics_mode));
  map.set(24, state.noise_gate_release);
  map.set(25, state.noise_gate_sens);
  map.set(26, state.swell_in);
  map.set(27, state.user_release);
  map.set(28, state.balance_filter);
  map.set(29, state.swell_out);
  map.set(31, envelopeModeToCCValue(state.envelope_mode));
  map.set(32, state.shifty_mode);
  map.set(33, spreadRoutingToCCValue(state.spread_routing));
  map.set(52, state.ramp_bounce ? 127 : 0);
  map.set(61, state.dip_dynamics ? 127 : 0);
  map.set(62, state.dip_attack ? 127 : 0);
  map.set(63, state.dip_eq ? 127 : 0);
  map.set(64, state.dip_dry ? 127 : 0);
  map.set(65, state.dip_wet ? 127 : 0);
  map.set(66, state.dip_bounce ? 127 : 0);
  map.set(67, sweepDirectionToCCValue(state.dip_sweep));
  map.set(68, polarityToCCValue(state.dip_polarity));
  map.set(71, state.dip_miso ? 127 : 0);
  map.set(72, state.dip_spread ? 127 : 0);
  map.set(73, state.dip_latch ? 127 : 0);
  map.set(74, state.dip_sidechain ? 127 : 0);
  map.set(75, state.dip_noise_gate ? 127 : 0);
  map.set(76, state.dip_motion ? 127 : 0);
  map.set(77, state.dip_swell_aux ? 127 : 0);
  map.set(78, state.dip_dusty ? 127 : 0);
  map.set(100, state.expression);
  map.set(102, state.bypass ? 127 : 0);
  map.set(103, state.swell ? 127 : 0);
  map.set(104, state.alt_mode ? 127 : 0);
  map.set(105, state.swell_hold ? 127 : 0);
  map.set(106, state.dynamics_max ? 127 : 0);
  return map;
}

// ============================================================================
// CC → Parameter
// ============================================================================

export function ccToParameter(cc: number, value: number): CleanParameter | null {
  switch (cc) {
    case 14:  return { Dynamics: value };
    case 15:  return { Sensitivity: value };
    case 16:  return { Wet: value };
    case 17:  return { Attack: value };
    case 18:  return { Eq: value };
    case 19:  return { Dry: value };
    case 20:  return { RampSpeed: value };
    case 21:  return { ReleaseMode: releaseModeFromCCValue(value) };
    case 22:  return { EffectMode: effectModeFromCCValue(value) };
    case 23:  return { PhysicsMode: physicsModeFromCCValue(value) };
    case 24:  return { NoiseGateRelease: value };
    case 25:  return { NoiseGateSens: value };
    case 26:  return { SwellIn: value };
    case 27:  return { UserRelease: value };
    case 28:  return { BalanceFilter: value };
    case 29:  return { SwellOut: value };
    case 31:  return { EnvelopeMode: envelopeModeFromCCValue(value) };
    case 32:  return { ShiftyMode: value };
    case 33:  return { SpreadRouting: spreadRoutingFromCCValue(value) };
    case 52:  return { RampBounce: value >= 64 };
    case 61:  return { DipDynamics: value >= 64 };
    case 62:  return { DipAttack: value >= 64 };
    case 63:  return { DipEq: value >= 64 };
    case 64:  return { DipDry: value >= 64 };
    case 65:  return { DipWet: value >= 64 };
    case 66:  return { DipBounce: value >= 64 };
    case 67:  return { DipSweep: sweepDirectionFromCCValue(value) };
    case 68:  return { DipPolarity: polarityFromCCValue(value) };
    case 71:  return { DipMiso: value >= 64 };
    case 72:  return { DipSpread: value >= 64 };
    case 73:  return { DipLatch: value >= 64 };
    case 74:  return { DipSidechain: value >= 64 };
    case 75:  return { DipNoiseGate: value >= 64 };
    case 76:  return { DipMotion: value >= 64 };
    case 77:  return { DipSwellAux: value >= 64 };
    case 78:  return { DipDusty: value >= 64 };
    case 100: return { Expression: value };
    case 102: return { Bypass: value >= 64 };
    case 103: return { Swell: value >= 64 };
    case 104: return { AltMode: value >= 64 };
    case 105: return { SwellHold: value >= 64 };
    case 106: return { DynamicsMax: value >= 64 };
    case 111: return { PresetSave: value };
    default:  return null;
  }
}
